File: src/main/java/io/github/flegacy/flms/command/branch/WandCommandBranch.java
package io.github.flegacy.flms.command.branch;

import com.mojang.brigadier.Command;
import com.mojang.brigadier.builder.LiteralArgumentBuilder;
import io.github.flegacy.flms.FLMS;
import io.github.flegacy.flms.ItemLibrary;
import io.github.flegacy.flms.SoundLibrary;
import io.github.flegacy.flms.utils.TextConstants;
import io.papermc.paper.command.brigadier.CommandSourceStack;
import io.papermc.paper.command.brigadier.Commands;
import org.bukkit.entity.Player;

public class WandCommandBranch implements CommandBranch {

	private static final String BRANCH_LITERAL = "wand";

	private final ItemLibrary itemLibrary;
	private final SoundLibrary soundLibrary;

	public WandCommandBranch(FLMS plugin) {
		itemLibrary = plugin.getItemLibrary();
		soundLibrary = plugin.getSoundLibrary();
	}

	@Override
	public LiteralArgumentBuilder<CommandSourceStack> getCommandTree() {
		return Commands.literal(BRANCH_LITERAL)
				.executes(ctk -> givePlayerWand(ctk.getSource()));
	}

	private int givePlayerWand(CommandSourceStack source) {
		if (!(source.getSender() instanceof Player player)) {
			source.getSender().sendMessage(TextConstants.errorMessage(TextConstants.ERROR_COMMAND_CONSOLE));
			return 0;
		}
		if (player.getInventory().firstEmpty() != -1) {
			player.getInventory().addItem(itemLibrary.getWand());
			player.sendMessage(TextConstants.prefixedMessage("Right click with the wand to use it!"));
			soundLibrary.playItemPickupSound(player);
			return Command.SINGLE_SUCCESS;
		}
		else {
			player.sendMessage(TextConstants.errorMessage(TextConstants.ERROR_INVENTORY_FULL));
			soundLibrary.playErrorSound(player);
			return 0;
		}
	}
}

File: src/main/kotlin/io/github/flegacy/flms/command/branch/WandCommandBranch.kt
package io.github.flegacy.flms.command.branch

import com.mojang.brigadier.Command
import com.mojang.brigadier.builder.LiteralArgumentBuilder
import io.github.flegacy.flms.FLMS
import io.github.flegacy.flms.util.ERROR_COMMAND_CONSOLE
import io.github.flegacy.flms.util.ERROR_INVENTORY_FULL
import io.github.flegacy.flms.util.errPrefixed
import io.github.flegacy.flms.util.prefixed
import io.github.flegacy.flms.util.soundPickup
import io.papermc.paper.command.brigadier.CommandSourceStack
import io.papermc.paper.command.brigadier.Commands
import org.bukkit.entity.Player

private const val BRANCH_LITERAL = "wand"

class WandCommandBranch(private val plugin: FLMS): CommandBranch {

    override fun buildCommandTree(): LiteralArgumentBuilder<CommandSourceStack> {
        return Commands.literal(BRANCH_LITERAL)
            .executes { ctx -> execute(ctx.source) }
    }

    private fun execute(source: CommandSourceStack): Int {
        if (source.sender !is Player) {
            source.sender.sendMessage(errPrefixed(ERROR_COMMAND_CONSOLE))
            return 0
        }

        val player = source.sender as Player

        if (player.inventory.firstEmpty() == -1) {
            player.sendMessage(errPrefixed(ERROR_INVENTORY_FULL))
            return 0
        }

        soundPickup(player)
        player.inventory.addItem(plugin.itemLib().wand())
        player.sendMessage(prefixed("The FLMS wand was added to your inventory."))
        return Command.SINGLE_SUCCESS
    }

}
